L44-python-rich-status.md

```diff
@@ -0,0 +1,394 @@
+
+### Client side API
+
+Currently, the documentation about getting status resides in [grpc.Call](https://grpc.io/grpc/python/grpc.html#client-side-context) section. We propose to add a new member function named `status()`, it will return a `namedtuple('Status', ['code', 'message', 'details'])`.
+
+The usage snippet will look like:
+
+```Python
+stub = ...Stub(channel)
+
+try:
+    resp = stub.AMethodHandler(...)
+except grpc.RpcError as rpc_error:
+    code = rpc_error.status().code
+    message = rpc_error.status().message
+    details = rpc_error.status().details
+    # or
+    code, message, details = rpc_error.status()
+```
+
+### Server side API
+
+gRPC Python team encountered lots of disagreements about how this part should be implemented. Available options will be listed in the `Rationale` section. Here I will post the final consensus for the team, but feel free to comment about other options.
+
+The new API will be added to [ServicerContext](https://grpc.io/grpc/python/grpc.html#service-side-context), and named `set_status`. It accept two positional arguments and a keyword argument, so setting the status code and status message are mandatory but not details. At the same time, current API `set_code`/`set_details` will be labeled as "deprecated".
+
+The function signature will be:
+
+```Python
+def set_status(code, message, details=""): pass
+```
+
+The usage will look like:
+
+```Python
+def ...Servicer(...):
+    def ARPCCall(request, context):
+        ...
+        context.set_status(
+            grpc.StatusCode.INVALID_ARGUMENT,
+            "Failed to recognize input variables",
+        )
+```
+
+### Rich Status Usage
+
+Besides this proposal, there will be a new gRPC Python extension package that depends on ProtoBuf named `grpcio-status`. The new package will provide convenient functions to help developers transform from ProtoBuf instance to gRPC status. The usage will look like:
+
+```Python
+# Client side
+from grpc_status import rpc_status
+
+stub = ...Stub(channel)
+
+try:
+    resp = stub.AMethodHandler(...)
+except grpc.RpcError as rpc_error:
+    rich_status = rpc_status.from_rpc_error(rpc_error)
+    # `rich_status` here is a ProtoBuf instance of `google.rpc.status.Status` proto message
+```
+```Python
+# Server side
+from grpc_status import rpc_status
+from google.protobuf import any_pb2
+
+def ...Servicer(...):
+    def ARPCCall(request, context):
+        ...
+        detail = any_pb2.Any()
+        detail.Pack(
+            rpc_status.error_details_pb2.DebugInfo(
+                stack_entries=traceback.format_stack(),
+                detail="Can't recognize this argument",
+            )
+        )
+        rich_status = grpc_status.status_pb2.Status(
+            code=grpc_status.code_pb2.INVALID_ARGUMENT,
+            message='API call quota depleted',
+            details=[detail]
+        )
+
+        context.set_status(*rpc_status.convert(rich_status))
```

Comment:
It's a little bit odd to have to use a splat here. Any way we can eliminate it?

Reply:
I don't have better idea than this one. If we are going to make two variable mandatory and one optional, this is my best solution. Also, I don't like to make a function accept different set of arguments.

An alternative way can satisfy both criteria is make the function signature like this:
```Python
def set_status(self, status_tuple):
```
```Python
# Usage for non-grpcio-status users

context.set_status((code, message))
```
But the additional parentheses doesn't look good.


Do you have any insight?

Reply:
My thought was to make the function accept different types of arguments and determine at runtime how to unpack them. But there are certainly arguments against that type of approach. The splat should be fine.